### DISCUSSION TITLE
Add dynamic modifierStyles to react-day-picker

### DISCUSSION THREAD

**Discussion by RageCrezz** [gpbl/react-day-picker#2086] [Support]
Hello,

I'm relatively new to contributing on GitHub, so I appreciate your patience. 😄 
While browsing through the issue list, I stumbled upon a solution for adding an array of ranges to the react-day-picker library. Below is the current code snippet I've been working with:

```tsx
import { Calendar as CalendarComp } from '@/components/ui/calendar'
import {
  DateRange,
  InternalModifier,
  InternalModifiers,
} from 'react-day-picker'
import CardView from '../Card/Card'
import { CalendarProps } from './Calendar.types'

const SELECTED_DATES: DateRange[] = [
  {
    from: new Date(2024, 3, 11),
    to: new Date(2024, 3, 14),
    bgColor: '#E74C3C',
  },
  {
    from: new Date(2024, 3, 15),
    to: new Date(2024, 3, 19),
    bgColor: '#2ECC71',
  },
]

function dateRangeModifiers(ranges: DateRange[]) {
  return ranges.reduce<
    Pick<
      InternalModifiers,
      | InternalModifier.RangeEnd
      | InternalModifier.RangeMiddle
      | InternalModifier.RangeStart
    >
  >(
    (prev, curr) => ({
      range_start: curr.from
        ? [...prev.range_start, curr.from]
        : prev.range_start,
      range_end: curr.to ? [...prev.range_end, curr.to] : prev.range_end,
      range_middle:
        curr.from && curr.to
          ? [...prev.range_middle, { after: curr.from, before: curr.to }]
          : prev.range_middle,
    }),
    {
      range_start: [],
      range_end: [],
      range_middle: [],
    },
  )
}

const Calendar = ({
  selected,
  onSelect,
  dateRange,
  ...rest
}: CalendarProps): JSX.Element => {
  const ranges: DateRange[] = SELECTED_DATES.map(s => ({
    from: s.from,
    to: s.to,
    bgColor: s.bgColor,
  }))

  return (
    <CardView>
      <CalendarComp
        mode="range"
        selected={ranges}
        onSelect={onSelect}
        modifiers={dateRangeModifiers(ranges)}
        {...rest}
      />
    </CardView>
  )
}

export default Calendar
```

As you can see, I've modified the DateRange type to include a `bgColor` property. My objective is to leverage this `bgColor` property to dynamically generate `modifierStyles` for `range_start` and `range_end`. Ultimately, I'm aiming to achieve different background colors for the start and end of each date range.

Any insights or suggestions on how to implement this would be greatly appreciated!

Thank you in advance for your assistance!

**Comment by gpbl** [comment]
Hi @RageCrezz, I understand you want to display multiple ranges in DayPicker with `mode="range"`. Not an easy task: it could get a bit tricky to solve this and I hope it will improve in the next version.

Currently multiple ranges is not supported by the "range" mode, but you can implement a similar behavior toggling off the selection mode, and implement one by yourself.

### Switching to default mode , add onDayClick

```diff
<CalendarComp
-  mode="range"
+  mode="default"
   selected={ranges}
-  onSelect={onSelect}
+  onDayClick={doSomethingWithTheClickedDay}
   modifiers={dateRangeModifiers(ranges)}
   {...rest}
/>
```

### Changing the background color 

If you want to change the background color for each range, using `modifiersStyles` won't help you: these are applied to the modifiers with the same name, regardless the range they belong to.

You can however use a different component to render the day cell. Find the matching range and its color and override the background color of the `button` in case is selected.

#### 1. add a new component called DayWithColor:

```tsx
/** Returns the bgColor for the given date. */
function findColorForDate(date: Date) {
  return selectedRanges.find((range) => {
    return range.from && range.from <= date && range.to && date <= range.to;
  })?.bgColor;
}

/**
 * The content of a day cell – will override the background color if the day is in a range.
 * This basically rewrites https://github.com/gpbl/react-day-picker/blob/main/src/components/Day/Day.tsx
 */
export function DayWithColor(props: DayProps): JSX.Element {
  const buttonRef = useRef<HTMLButtonElement>(null);

  const { isHidden, isButton, divProps, buttonProps } = useDayRender(
    props.date,
    props.displayMonth,
    buttonRef
  );

  const dateColor = findColorForDate(props.date);

  if (isHidden) {
    return <div role="gridcell"></div>;
  }
  if (!isButton) {
    const divStyle = {
      ...divProps.style,
      backgroundColor: dateColor ?? divProps.style?.backgroundColor
    };
    return <div {...divProps} style={divStyle} />;
  }
  const buttonStyle = {
    ...buttonProps.style,
    backgroundColor: dateColor ?? buttonProps.style?.backgroundColor
  };
  return (
    <Button name="day" ref={buttonRef} {...buttonProps} style={buttonStyle} />
  );
}
```

#### 2. Adjust DayPicker to use this component as `Day`: pass it to the `components` prop.

```tsx
export default function App() {
  return (
    <DayPicker
      components={{ Day: DayWithColor }} // Replace the `Day` component with yours
      mode="default" // make sure you are in default selection mode
      selected={selectedRanges} // mark the days as selected 
      onDayClick={(date: Date) => console.log(date)} // you have to implement your own selection mode
      modifiers={dateRangeModifiers(selectedRanges)} // add range modifiers (from/to/middle)
    />
  );
}
```

See this code sandbox with the example above:

- Preview: https://ycndkk-5173.csb.app
- Source: https://codesandbox.io/p/devbox/ycndkk?file=%2Fsrc%2FApp.tsx%3A22%2C33

<img width="908" alt="Screenshot 2024-04-10 at 9 45 10 AM" src="https://github.com/gpbl/react-day-picker/assets/120693/1cf0ae5d-a0c5-4c0d-a2b2-8aa2313d7c72">


Let me know how it goes and good luck!

**Reply by MathiasPawshake** [reply]
Hi @gpbl, would be nice to have multiple ranges inside v9!

I tried to access the codesandbox but I can't access it because of permissions. Was this project removed?

**Reply by gpbl** [reply]
> Hi @gpbl, would be nice to have multiple ranges inside v9!

Some requirements could help - how should it work? 

> I tried to access the codesandbox but I can't access it because of permissions. Was this project removed?

Wrong permissions - sorry. https://codesandbox.io/p/devbox/ycndkk

**Reply by MathiasPawshake** [reply]
Basicly it works the same as range mode but you can just select multiple ranges instead of just being able to select one range
